### PR TITLE
KAFKA-5970: Use ReentrantLock for delayed operation lock to avoid blocking

### PR DESCRIPTION
**Pull request by rajinisivaram**

### CODE REVIEW

**Comment by rajinisivaram** [comment]
@hachikuji Can you take a look to see if this approach makes sense to reduce locking? I haven't yet looked at how to unit test this. Thank you!

**Comment by hachikuji** [comment]
@rajinisivaram Thanks for the patch and nice catch! If you look at the code in `GroupMetadataManager`, we use the group itself as the lock when appending to the log through `ReplicaManager`: https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala#L247. What I'm wondering is whether we should do the same thing when writing the transaction markers to __consumer_offsets. In other words, maybe the markers should be written to the log from the context of `GroupMetadataManager` when the target topic is __consumer_offsets?  This would add a bit of complexity to the `WriteTxnMarker` handler, but it might be more consistent and easier to understand. What do you think?

**Comment by rajinisivaram** [comment]
@hachikuji Thank you for the review. Yes, I had seen the group being used as the lock in `GroupMetadataManager` and from a consistency point of view, it does make sense to do the same for `WriteTxnMarker` too. But just to clarify - that in itself is not sufficient to avoid deadlocks, right? 

I am struggling to understand why a deadlock (without transactions) was not observed before - I don't mean in tests where it is rather unlikely, but by users. Though this deadlock was due to transactions, if we hold a group lock while appending to logs, even the process of checking whether `DelayedProduce` of another group on the same partition is ready to be completed could acquire a different group lock. And holding two group locks in arbitrary order is bound to lead to deadlocks. Have I missed something?

**Comment by hachikuji** [comment]
We've had multiple deadlock problems in the past due to locks being held when the delayed operation was executed. From memory, one problem was inside `DelayedOperationPurgatory` itself, but we changed the internal data structure so that we didn't need to hold a lock while iterating it. There was another one where a lock was being held in `ReplicaManager` when the delayed operation was executed. Finally, what we thought was the last problem: the lock on the `DelayedOperation` instance itself. We protected all of the operations in `GroupMetadataManager` from this deadlock potential by allowing the delayed operation to lock on a separate object than itself (in this case, the group). But we missed this case where we needed to write transaction markers.

In general, we have found it difficult to restructure the `GroupCoordinator` logic to avoid holding the group lock when writing to the log or completing delayed operations. Giving up the lock means opening the door to another request from the same group coming in and altering the state in some way. It has been easier to try and restructure the purgatory code to avoid holding unneeded locks when invoking delayed operation, though this has probably resulted in more missed cases. We faced similar problems inside the `TransactionCoordinator` and ultimately solved them with the same solution.

**Comment by rajinisivaram** [comment]
@hachikuji Thanks for the explanation.  I do agree that if `WriteTxnMarker` can be fixed in the same way as the other writes, it makes sense to do so, without changing the current level of locking, so that we don't reduce synchronization and introduce new bugs in the process. But I am not sure how that is possible without further changes to the purgatory code.

When I saw the stack trace in the failure with `WriteTxnMarker`, what surprised me was not that the `WriteTxnMarker` led to a deadlock, since that is relatively new code. It was the fact that the stack trace showed that we were holding a group lock while doing an append.  And appends attempt to lock all other groups in the purgatory for the same partition. To me that looks like a recipe for deadlocks. But I thought I must be missing something since no one has reported a deadlock for this scenario. So I was kind of relieved to see the deadlock that Peter Davis reported on the JIRA because that contains exactly the stack trace that I was expecting with the current code, without transactions. It seems to me that if we really want to keep the current level of synchronization in `GroupCoordinator`, it would be better to use read/write locks for the purgatory. But I am not familiar with the code or with the deadlock fixes that went in earlier, so I could just have missed something.

Perhaps we should start with the other stack trace with offset commit, make sure that is fixed first and then apply the same approach for transaction markers?

**Comment by ijuma** [comment]
I agree that working on a fix that fixes the non transactional case first makes sense.

File: core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala [hunk 1]
```diff
@@ -765,20 +777,26 @@ class GroupCoordinator(val brokerId: Int,
...
+      shouldKeepMemberAlive(member, heartbeatDeadline) || member.isLeaving
     }
+    if (complete)
+      forceComplete()
+    else
+      false
   }
 
   def onExpireHeartbeat(group: GroupMetadata, member: MemberMetadata, heartbeatDeadline: Long) {
+    var preparingRebalance = false
```
Comment (ijuma) [inline core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala]: I wonder if we should introduce a helper method that runs another function outside the lock based on the result of the original function. Not sure if useful, just an idea.

**Comment by rajinisivaram** [comment]
@hachikuji @ijuma I tried out two PRs. In the first one, I retained the same locking design as now, with group locks and transaction metadata locks turned into a `ReentrantLock`. But I think that approach doesn't help with `GroupMetadataManager.handleTxnCompletion` which locks multiple groups and hence cannot override the lock in `DelayedProduce`. So I went for a second approach, which is the commit in this PR.

All delayed operations are handled in exactly the same way. It simplifies the code, but it does mean that we create a `ReentrantLock` for every operation that is delayed and needs to be watched. Groups continue to synchronize on the `GroupMetadata` object, but create `DelayedProduce` with its own `ReentrantLock`. Attempts to complete delayed operations are made only if the delayed lock can be acquired without blocking. Let me know if you think this approach is too expensive or flawed. Thank you!

**Comment by hachikuji** [comment]
@rajinisivaram Thanks for the update. Your approach is interesting. We allow the locks to be acquired in different orders, but we break the deadlock by making the acquisition of the operation lock optional. So even if one thread has acquired the group lock and another thread has acquired the operation lock, only the thread with the operation lock will be blocked. The other thread will continue and eventually give up the group lock. Is that right?

**Comment by rajinisivaram** [comment]
@hachikuji Yes, that is exactly right.

**Comment by hachikuji** [comment]
cc @guozhangwang It seems like this works, but would be nice to get your thoughts as well.

**Comment by guozhangwang** [comment]
Checked all callers of `safeTryComplete` to make sure that there is no case that if we skip one operation due to lock being held but completion criterion not being satisfied yet, then it will never be able to be completed any more. Seems all good to me. cc @rajinisivaram @hachikuji

File: core/src/main/scala/kafka/server/DelayedOperation.scala [hunk 2]
```diff
@@ -96,13 +101,18 @@ abstract class DelayedOperation(override val delayMs: Long) extends TimerTask wi
   def tryComplete(): Boolean
 
   /**
-   * Thread-safe variant of tryComplete(). This can be overridden if the operation provides its
-   * own synchronization.
+   * Thread-safe variant of tryComplete() that attempts completion only if the lock can be acquired
+   * without blocking.
    */
   def safeTryComplete(): Boolean = {
```
Comment (hachikuji) [inline core/src/main/scala/kafka/server/DelayedOperation.scala]: If we don't need the overrides anymore, can this be private? Also, how about `maybeTryComplete`?

File: core/src/main/scala/kafka/server/DelayedOperation.scala [hunk 3]
```diff
@@ -46,6 +46,11 @@ import scala.collection.mutable.ListBuffer
 abstract class DelayedOperation(override val delayMs: Long) extends TimerTask with Logging {
 
   private val completed = new AtomicBoolean(false)
+  protected var lock: ReentrantLock = null
+
+  def createLock(): Unit = {
+    lock = new ReentrantLock
```
Comment (ijuma) [inline core/src/main/scala/kafka/server/DelayedOperation.scala]: Is the lazy creation worth it? I'd imagine that the other state in the operation would be much larger than the lock (we typically have a map of partitions to some value). If we decide that keeping it lazy is worth it, don't we have to make it `volatile`?

File: core/src/main/scala/kafka/server/DelayedOperation.scala [hunk 4]
```diff
@@ -19,11 +19,11 @@ package kafka.server
 
 import java.util.concurrent._
 import java.util.concurrent.atomic._
-import java.util.concurrent.locks.ReentrantReadWriteLock
+import java.util.concurrent.locks.{ReentrantLock, ReentrantReadWriteLock}
 
 import com.yammer.metrics.core.Gauge
 import kafka.metrics.KafkaMetricsGroup
-import kafka.utils.CoreUtils.{inReadLock, inWriteLock}
+import kafka.utils.CoreUtils.{inLock, inReadLock, inWriteLock}
```
Comment (hachikuji) [inline core/src/main/scala/kafka/server/DelayedOperation.scala]: Unused `inLock`?

**Review by hachikuji — APPROVED** [review]
Minor comments, but LGTM.

File: core/src/main/scala/kafka/server/DelayedOperation.scala (hunk 3 above)
Comment (ijuma) [reply core/src/main/scala/kafka/server/DelayedOperation.scala]: Also, if we keep this approach, we should tighten the access modifiers. `createLock` doesn't need to be public, we should make the `var` private and expose a protected `lock` accessor, etc.

File: core/src/main/scala/kafka/server/DelayedOperation.scala (hunk 2 above)
Comment (hachikuji) [reply core/src/main/scala/kafka/server/DelayedOperation.scala]: My mistake. It does need to be exposed.

File: core/src/main/scala/kafka/server/DelayedOperation.scala (hunk 3 above)
Comment (rajinisivaram) [reply core/src/main/scala/kafka/server/DelayedOperation.scala]: Removed lazy initialization.

File: core/src/main/scala/kafka/server/DelayedOperation.scala (hunk 2 above)
Comment (rajinisivaram) [reply core/src/main/scala/kafka/server/DelayedOperation.scala]: Made it private in `server` and renamed to `maybeTryComplete`.

**Comment by hachikuji** [comment]
The one failure seems unrelated. Merging to trunk.

**Pull request closed without merging**